lib/stm32/common/spi_common_v3.c

```diff
@@ -0,0 +1,834 @@
+		SPI_IFCR(spi) = SPI_IFCR_EOTC;
+	}
+	SPI_CR2(spi) = (SPI_CR2(spi) & ~(SPI_CR2_TSIZE_MASK << SPI_CR2_TSIZE_SHIFT))
+		| (size << SPI_CR2_TSIZE_SHIFT);
+
+	SPI_CR1(spi) |= SPI_CR1_CSTART;
+}
+
+void spi_wait_for_transfer_finished(uint32_t spi)
+{
+	while ((SPI_SR(spi) & SPI_SR_EOT) == 0);
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief SPI Set Full Duplex (3-wire) Mode
+
+@param[in] spi Unsigned int32. SPI peripheral identifier @ref spi_reg_base.
+*/
+
+void spi_set_full_duplex_mode(uint32_t spi)
+{
+	SPI_CFG2(spi) = (SPI_CFG2(spi) & ~(SPI_CFG2_COMM_MASK << SPI_CFG2_COMM_SHIFT))
+		| (SPI_CFG2_COMM_FULL_DUPLEX << SPI_CFG2_COMM_SHIFT);
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief SPI Set Receive Only Mode for Simplex (2-wire) Unidirectional
+Transfers
+
+@param[in] spi Unsigned int32. SPI peripheral identifier @ref spi_reg_base.
+*/
+
+void spi_set_receive_only_mode(uint32_t spi)
+{
+	SPI_CFG2(spi) = (SPI_CFG2(spi) & ~(SPI_CFG2_COMM_MASK << SPI_CFG2_COMM_SHIFT))
+		| (SPI_CFG2_COMM_SIMPLEX_RX << SPI_CFG2_COMM_SHIFT);
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief SPI Disable Slave Management by Hardware
+
+In slave mode the NSS hardware input is used as a select enable for the slave.
+
+@param[in] spi Unsigned int32. SPI peripheral identifier @ref spi_reg_base.
+*/
+
+void spi_disable_software_slave_management(uint32_t spi)
+{
+	SPI_CFG2(spi) &= ~SPI_CFG2_SSM;
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief SPI Enable Slave Management by Software
+
+In slave mode the NSS hardware input is replaced by an internal software
+enable/disable of the slave (@ref spi_set_nss_high).
+
+@param[in] spi Unsigned int32. SPI peripheral identifier @ref spi_reg_base.
+*/
+
+void spi_enable_software_slave_management(uint32_t spi)
+{
+	SPI_CFG2(spi) |= SPI_CFG2_SSM;
+	/* allow slave select to be an input */
+	SPI_CFG2(spi) |= SPI_CFG2_SSOE;
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief SPI Set the Software NSS Signal High
+
+In slave mode, and only when software slave management is used, this replaces
+the NSS signal with a slave select enable signal.
+
+@todo these should perhaps be combined with an SSM enable as it is meaningless
+otherwise
+
+@param[in] spi Unsigned int32. SPI peripheral identifier @ref spi_reg_base.
+*/
+
+void spi_set_nss_high(uint32_t spi)
+{
+	SPI_CR1(spi) |= SPI_CR1_SSI;
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief SPI Set the Software NSS Signal Low
+
+In slave mode, and only when software slave management is used, this replaces
+the NSS signal with a slave select disable signal.
+
+@param[in] spi Unsigned int32. SPI peripheral identifier @ref spi_reg_base.
+*/
+
+void spi_set_nss_low(uint32_t spi)
+{
+	SPI_CR1(spi) &= ~SPI_CR1_SSI;
+}
```

Comment:
cmon, this is all in spi_common_all.  Why can't we extend that? Are there some "alL" functions that are not "v12"?  then we just do that.  We'v edone that before, we'll do it again..